File: testing/c/regex/dfa0.nim
import std/[algorithm]
import hmisc/hdebug_misc

proc re2post(inRe: string): string =
  var
    nalt: int
    natom: int
    dst: int

  var paren: array[100, tuple[nalt, natom: int]]
  var p: int

  for idx, re in pairs(inRe):
    case re:
      of '(':
        if (natom > 1):
          dec natom
          result.add '.'

        # if (p >= paren + 100) {return NULL;}

        paren[p].nalt  = nalt
        paren[p].natom = natom
        inc p
        nalt  = 0
        natom = 0

      of '|':
        # if (natom == 0) {return NULL;} # FIXME

        while ((dec natom; natom) > 0):
          result.add '.'

        inc nalt

      of ')':
        # if (p == paren) {return NULL;}
        # if (natom == 0) {return NULL;}

        while ((dec natom; natom) > 0):
          result.add '.'

        while nalt > 0:
          result.add '|'
          dec nalt

        dec p

        nalt  = paren[p].nalt
        natom = paren[p].natom
        inc natom

      of '*', '+', '?':
        # if (natom == 0) {return NULL;}
        result.add re

      else:
        if (natom > 1):
          dec natom;
          result.add '.'

        result.add re
        inc natom

    # if (p != paren) {return NULL;}

    while natom > 0:
        result.add '.'
        dec natom

    while nalt > 0:
      result.add '|'
      dec nalt

type
  StateKind = enum
    skValue
    skMatch
    skSplit

  State = ref object
    lastlist: int
    value: char
    out0: State
    out1: State
    kind: StateKind

  Frag = ref object
    start: State
    outList: PtrList

  PtrList {.union.} = ref object
    next: PtrList
    state: State

  List = ref seq[State]
  DState = ref object
    l: List
    next: array[char, DState]
    left, right: int




var nstate: int

proc newState(out0, out1: State): State =
  result = State(kind: skSplit, out0: out0, out1: out1, lastlist: 0)
  inc nstate

proc newState(ch: char): State =
  result = State(kind: skValue, value: ch)
  inc nstate

proc newFrag(start: State, outList: Ptrlist): Frag =
  Frag(start: start, outList: outList)

proc newList(state: State): PtrList =
  PtrList(next: nil, state: state)

proc patch(l: Ptrlist, s: State) =
  var next: PtrList
  var l = l
  while not isNil(l):
    next = l.next
    l.state = s
    l = next

proc append(l1, l2: Ptrlist): Ptrlist =
  var oldl1 = l1
  var l1 = l1

  while not isNil(l1.next):
    l1 = l1.next

  l1.next = l2
  return oldl1

proc post2nfa(postfix: string): State =
  var stack: seq[Frag]
  # char*  p;
  var e1, e2, e: Frag
  var s: State

  for p in postfix:
    case p:
      of '.':
        e2 = stack.pop()
        e1 = stack.pop()
        patch(e1.outList, e2.start)
        stack.add(newFrag(e1.start, e2.outList))

      of '|':
        e2 = stack.pop()
        e1 = stack.pop()
        s  = newState(e1.start, e2.start)
        stack.add(newFrag(s, append(e1.outList, e2.outList)))

      of '?':
        e = stack.pop()
        s = newState(e.start, nil)
        stack.add(newFrag(s, append(e.outList, newList(s.out1))))

      of '*':
        e = stack.pop()
        s = newState(e.start, nil)
        patch(e.outList, s)
        stack.add(newFrag(s, newList(s.out1)))

      of '+':
        e = stack.pop()
        s = newState(e.start, nil)
        patch(e.outList, s)
        stack.add(newFrag(e.start, newList(s.out1)))

      else:
        s = newState(p)
        stack.add(newFrag(s, newList(s.out0)))



    e = stack.pop()
    # if (stackp != stack):
    #     return nil

    patch(e.outList, State(kind: skMatch));
    return e.start

var l1, l2: List
var listid: int

iterator items(list: List): auto =
  for item in list[]:
    yield item

proc add(list: List, item: State) = list[].add item
proc len(list: List): int = list[].len

proc ismatch(l: List): bool =
  for state in l:
    if state.kind == skMatch:
      return true

  return false

proc addstate(l: var List, s: State) =
    if (isNil(s) or s.lastlist == listid):
      return

    s.lastlist = listid
    if (s.kind == skSplit):
        addstate(l, s.out0)
        addstate(l, s.out1)
        return

    l.add s

proc startlist(start: var State, l: var List): List =
  inc listid
  addstate(l, start)
  return l



proc step(clist: var List, c: char, nlist: var List) =
  inc listid
  for s in clist:
    if s.kind == skValue and s.value == c:
      addstate(nlist, s.out0)


proc listcmp(l1, l2: List): int =
  if (l1.len < l2.len):
    return -1

  if (l1.len > l2.len):
    return 1

  for i in 0 ..< l1.len:
    if (l1[i] < l2[i]):
      return -1

    elif (l1[i] > l2[i]):
      return 1

  return 0

proc ptrcmp(a, b: ref): int =
  if (a < b):
    return -1

  elif (a > b):
    return 1

  else:
    return 0

var alldstates: seq[DState]
proc dstate(l: List): DState =
  sort(l[]) do (s1, s2: State) -> int:
    ptrcmp(s1, s2)

  var dp = 0
  var d: DState = alldstates[dp]
  echo dp, " -> "
  while not isNil(d):
    # FIXME I don't understand original C code good enough to rewrite it
    # into something same. Specifically, article says we are iterating over
    # binary tree here - that part is easy, but I literally can not find
    # any other place where `.left` and `.right` is referenced, which means
    # they are either not set at all (highly unlikely), or this is done via
    # some badshit-crazy C memory hacks, that I can'd decode.
    let i = listcmp(l, d.l)
    if i < 0:
      dp = d.left

    elif i > 0:
      dp = d.right

    else:
      echo "Found value"
      return d

    d = alldstates[dp]

  d = DState()
  new(d.l)

  alldstates[dp] = d
  return d



#   dp = &alldstates;
#   while ((d = *dp) != NULL):
#     let i = listcmp(l, &d->l);
#     if (i < 0):
#       dp = &d->left;

#     elif (i > 0):
#       dp = &d->right;

#     else:
#       return d

#   d = malloc(sizeof *d + l->n * sizeof l->s[0]);
#   memset(d, 0, sizeof *d);
#   d->l.s = (State**)(d + 1);
#   memmove(d->l.s, l->s, l->n * sizeof l->s[0]);
#   d->l.n = l->n;
#   *dp    = d;
#   return d;

proc startnfa(start: State, l: var List) =
  inc listid
  addstate(l, start);

proc startdstate(start: var State): DState =
  return dstate(startlist(start, l1))


proc nextstate(d: DState, c: char): DState =
  step(d.l, c, l1)
  d.next[c] = dstate(l1)
  return d.next[c]

proc match(start: DState, s: string): bool =
  var
    d: DState = start
    next: DState

  for c in s:
    next = d.next[c]
    if isNIl(next):
      next = nextstate(d, c)

    d = next

  return ismatch(d.l)

proc main() =
  var post  = re2post("a*b");
  echo post
  var start = post2nfa(post);
  echo isNil(start)
  new(l1); l1[] = newSeq[State](nstate)
  new(l2); l2[] = newSeq[State](nstate)
  alldstates = newSeq[DState](nstate)
  # l1.s         = malloc(nstate * sizeof l1.s[0]);
  # l2.s         = malloc(nstate * sizeof l2.s[0]);

  startHax()
  let str = "aaaab"
  echov match(startdstate(start), str)

main()

File: testing/c/regex/dfa0.c
/*
 * Regular expression implementation.
 * Supports only ( | ) * + ?.  No escapes.
 * Compiles to NFA and then simulates NFA
 * using Thompson's algorithm.
 * Caches steps of Thompson's algorithm to
 * build DFA on the fly, as in Aho's egrep.
 *
 * See also http://swtch.com/~rsc/regexp/ and
 * Thompson, Ken.  Regular Expression Search Algorithm,
 * Communications of the ACM 11(6) (June 1968), pp. 419-422.
 *
 * Copyright (c) 2007 Russ Cox.
 * Can be distributed under the MIT license, see bottom of file.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>


/*
 * Convert infix regexp re to postfix notation.
 * Insert . as explicit concatenation operator.
 * Cheesy parser, return static buffer.
 */
char* re2post(char* re) {
    int         nalt, natom;
    static char buf[8000];
    char*       dst;
    struct {
        int nalt;
        int natom;
    } paren[100], *p;

    p     = paren;
    dst   = buf;
    nalt  = 0;
    natom = 0;
    if (strlen(re) >= sizeof buf / 2) {
        return NULL;
    }
    for (; *re; re++) {
        switch (*re) {
            case '(': {
                if (natom > 1) {
                    --natom;
                    *dst++ = '.';
                }
                if (p >= paren + 100) {
                    return NULL;
                }
                p->nalt  = nalt;
                p->natom = natom;
                p++;
                nalt  = 0;
                natom = 0;
                break;
            }
            case '|': {
                if (natom == 0) {
                    return NULL;
                }
                while (--natom > 0) {
                    *dst++ = '.';
                }
                nalt++;
                break;
            }
            case ')':
                if (p == paren) {
                    return NULL;
                }
                if (natom == 0) {
                    return NULL;
                }
                while (--natom > 0) {
                    *dst++ = '.';
                }
                for (; nalt > 0; nalt--) {
                    *dst++ = '|';
                }
                --p;
                nalt  = p->nalt;
                natom = p->natom;
                natom++;
                break;
            case '*':
            case '+':
            case '?':
                if (natom == 0) {
                    return NULL;
                }
                *dst++ = *re;
                break;
            default:
                if (natom > 1) {
                    --natom;
                    *dst++ = '.';
                }
                *dst++ = *re;
                natom++;
                break;
        }
    }
    if (p != paren) {
        return NULL;
    }
    while (--natom > 0) {
        *dst++ = '.';
    }
    for (; nalt > 0; nalt--) {
        *dst++ = '|';
    }
    *dst = 0;
    return buf;
}

/*
 * Represents an NFA state plus zero or one or two arrows exiting.
 * if c == Match, no arrows out; matching state.
 * If c == Split, unlabeled arrows to out and out1 (if != NULL).
 * If c < 256, labeled arrow with character c to out.
 */
enum
{
    Match = 256,
    Split = 257
};
typedef struct State State;
struct State {
    int    code;
    State* out;
    State* out1;
    int    lastlist;
};
State matchstate = {Match}; /* matching state */
int   nstate;

/* Allocate and initialize State */
State* state(int c, State* out, State* out1) {
    State* s;

    nstate++;
    s           = malloc(sizeof *s);
    s->lastlist = 0;
    s->code     = c;
    s->out      = out;
    s->out1     = out1;
    return s;
}

/*
 * A partially built NFA without the matching state filled in.
 * Frag.start points at the start state.
 * Frag.out is a list of places that need to be set to the
 * next state for this fragment.
 */
typedef struct Frag   Frag;
typedef union Ptrlist Ptrlist;
struct Frag {
    State*   start;
    Ptrlist* out;
};

/* Initialize Frag struct. */
Frag frag(State* start, Ptrlist* out) {
    Frag n = {start, out};
    return n;
}

/*
 * Since the out pointers in the list are always
 * uninitialized, we use the pointers themselves
 * as storage for the Ptrlists.
 */
union Ptrlist {
    Ptrlist* next;
    State*   state;
};

/* Create singleton list containing just outp. */
Ptrlist* list1(State** outp) {
    Ptrlist* l;

    l       = (Ptrlist*)outp;
    l->next = NULL;
    return l;
}

/* Patch the list of states at out to point to start. */
void patch(Ptrlist* l, State* s) {
    Ptrlist* next;

    for (; l; l = next) {
        next     = l->next;
        l->state = s;
    }
}

/* Join the two lists l1 and l2, returning the combination. */
Ptrlist* append(Ptrlist* l1, Ptrlist* l2) {
    Ptrlist* oldl1;

    oldl1 = l1;
    while (l1->next) {
        l1 = l1->next;
    }
    l1->next = l2;
    return oldl1;
}

/*
 * Convert postfix regular expression to NFA.
 * Return start state.
 */
State* post2nfa(char* postfix) {
    char*  p;
    Frag   stack[1000], *stackp, e1, e2, e;
    State* s;

    // fprintf(stderr, "postfix: %s\n", postfix);

    if (postfix == NULL)
        return NULL;

#define push(s) *stackp++ = s
#define pop() *--stackp

    stackp = stack;
    for (p = postfix; *p; p++) {
        switch (*p) {
            default: {
                s = state(*p, NULL, NULL);
                push(frag(s, list1(&s->out)));
                break;
            }
            case '.': /* catenate */
            {
                e2 = pop();
                e1 = pop();
                patch(e1.out, e2.start);
                push(frag(e1.start, e2.out));
                break;
            }
            case '|': /* alternate */
            {
                e2 = pop();
                e1 = pop();
                s  = state(Split, e1.start, e2.start);
                push(frag(s, append(e1.out, e2.out)));
                break;
            }
            case '?': /* zero or one */
            {
                e = pop();
                s = state(Split, e.start, NULL);
                push(frag(s, append(e.out, list1(&s->out1))));
                break;
            }
            case '*': /* zero or more */
            {
                e = pop();
                s = state(Split, e.start, NULL);
                patch(e.out, s);
                push(frag(s, list1(&s->out1)));
                break;
            }
            case '+': /* one or more */
            {
                e = pop();
                s = state(Split, e.start, NULL);
                patch(e.out, s);
                push(frag(e.start, list1(&s->out1)));
                break;
            }
        }
    }

    e = pop();
    if (stackp != stack)
        return NULL;

    patch(e.out, &matchstate);
    return e.start;
#undef pop
#undef push
}

typedef struct List List;
struct List {
    State** s;
    int     n;
};
List       l1, l2;
static int listid;

void addstate(List*, State*);
void step(List*, int, List*);

/* Compute initial state list */
List* startlist(State* start, List* l) {
    l->n = 0;
    listid++;
    addstate(l, start);
    return l;
}

/* Check whether state list contains a match. */
int ismatch(List* l) {
    for (int i = 0; i < l->n; i++) {
        if (l->s[i] == &matchstate) {
            return 1;
        }
    }
    return 0;
}

/* Add s to l, following unlabeled arrows. */
void addstate(List* l, State* s) {
    if (s == NULL || s->lastlist == listid) {
        return;
    }
    s->lastlist = listid;
    if (s->code == Split) {
        /* follow unlabeled arrows */
        addstate(l, s->out);
        addstate(l, s->out1);
        return;
    }
    l->s[l->n++] = s;
}

/*
 * Step the NFA from the states in clist
 * past the character c,
 * to create next NFA state set nlist.
 */
void step(List* clist, int c, List* nlist) {
    int    i;
    State* s;

    listid++;
    nlist->n = 0;
    for (i = 0; i < clist->n; i++) {
        s = clist->s[i];
        if (s->code == c) {
            addstate(nlist, s->out);
        }
    }
}

/*
 * Represents a DFA state: a cached NFA state list.
 */
typedef struct DState DState;
struct DState {
    List    l;
    DState* next[256];
    DState* left;
    DState* right;
};

/* Compare lists: first by length, then by members. */
static int listcmp(List* l1, List* l2) {
    int i;

    if (l1->n < l2->n) {
        return -1;
    }
    if (l1->n > l2->n) {
        return 1;
    }
    for (i = 0; i < l1->n; i++) {
        if (l1->s[i] < l2->s[i]) {
            return -1;
        } else if (l1->s[i] > l2->s[i]) {
            return 1;
        }
    }
    return 0;
}

/* Compare pointers by address. */
static int ptrcmp(const void* a, const void* b) {
    if (a < b) {
        return -1;
    }
    if (a > b) {
        return 1;
    }
    return 0;
}

/*
 * Return the cached DState for list l,
 * creating a new one if needed.
 */
DState* alldstates;
DState* dstate(List* l) {
    int      i;
    DState **dp, *d;

    printf("alldstates %zu\n", alldstates);

    qsort(l->s, l->n, sizeof l->s[0], ptrcmp);
    dp = &alldstates;
    printf("%zu -> ", dp);
    while ((d = *dp) != NULL) {
        puts("  iteration");
        i = listcmp(l, &d->l);
        if (i < 0) {
            puts("Listcmp left");
            dp = &d->left;
        } else if (i > 0) {
            puts("Listcmp right");
            dp = &d->right;
        } else {
            puts("Found value");
            return d;
        }
    }
    puts("");

    d = malloc(sizeof *d + l->n * sizeof l->s[0]);
    printf("Create new state @%zu\n", d);
    memset(d, 0, sizeof *d);
    d->l.s = (State**)(d + 1);
    memmove(d->l.s, l->s, l->n * sizeof l->s[0]);
    d->l.n = l->n;
    puts("Assign to *dp");

    // Add value to list of all states. If original array pointer is zero
    // it will be set to some normal value
    *dp = d;
    printf("  alldstates %zu\n", alldstates);
    return d;
}

void startnfa(State* start, List* l) {
    l->n = 0;
    listid++;
    addstate(l, start);
}

DState* startdstate(State* start) {
    return dstate(startlist(start, &l1));
}

DState* nextstate(DState* d, int c) {
    step(&d->l, c, &l1);
    return d->next[c] = dstate(&l1);
}

/* Run DFA to determine whether it matches s. */
int match(DState* start, char* s) {
    DState *d, *next;
    d = start;
    for (; *s; s++) {
        int c = *s & 0xFF;
        if ((next = d->next[c]) == NULL) {
            next = nextstate(d, c);
        }
        d = next;
    }
    return ismatch(&d->l);
}

int main(int argc, char** argv) {
    char*  post  = re2post("a*b");
    State* start = post2nfa(post);
    l1.s         = malloc(nstate * sizeof l1.s[0]);
    l2.s         = malloc(nstate * sizeof l2.s[0]);

    puts("Started");
    char* str = "aaaab";
    if (match(startdstate(start), str)) {
        printf("%s\n", str);
    }

    return 0;
}


/*
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall
 * be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
